### PR TITLE
permissions: fix file permissions on intermediate install directories

### LINKED ISSUE #12102: spack creates app directory with permissions of 777
A fresh clone of spack is creating the install directory with permissions of 777. This began with commit 3a026f14123e8a010e0c969bb38dfdde2738226f, #11995. These are bad permissions, particularly when deploying a stack on a system.

With the following config:

```yaml
config:
  build_stage: /dev/shm
  misc_cache: $spack/var/spack/cache
  install_tree: $spack/../apps
```
an installation will create directories as follows:
```console
$ ls -l ../apps/                                                                        
total 0
drwxrwxrwx 3 gpjohnsn its-rs-user 31 Jul 23 15:12 linux-centos7-x86_64
$ ls -l ../apps/linux-centos7-x86_64/
total 0
drwxrwxrwx 3 gpjohnsn its-rs-user 66 Jul 23 15:12 gcc-4.8.5
```

The expected behavior is:
```console
$ ls -l ../apps/                                                                            
total 0
drwxr-xr-x 3 gpjohnsn its-rs-user 31 Jul 23 15:11 linux-centos7-x86_64
$ ls -l ../apps/                                                                            
total 0
drwxr-xr-x 3 gpjohnsn its-rs-user 31 Jul 23 15:11 linux-centos7-x86_64/gcc-4.8.5
```

### PR DESCRIPTION
**Pull request by becker33**
Fixes #12102 

Set file permissions for intermediate directories in install tree to the permissions of the parent directory.

### CODE REVIEW

File: lib/spack/llnl/util/filesystem.py [hunk 1]
```diff
@@ -226,7 +226,10 @@ def group_ids(uid=None):
 
 def chgrp(path, group):
     """Implement the bash chgrp function on a single path"""
-    gid = grp.getgrnam(group).gr_gid
+    if isinstance(group, six.string_type):
```
Comment (albestro) [inline lib/spack/llnl/util/filesystem.py]: I think it should be `six.string_types` (with s)

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -414,48 +417,84 @@ def get_filetype(path_name):
     return output.strip()
 
 
+def chgrp_if_not_world_writable(path, group):
+    """chgrp path to group if path is not world writable"""
+    mode = os.stat(path).st_mode
+    if mode & stat.S_IWOTH:
```
Comment (albestro) [inline lib/spack/llnl/util/filesystem.py]: From the function name it seems to me that you want to `chgrp` if _path_ is **NOT** writable, but from the check I would say the opposite.

File: lib/spack/llnl/util/filesystem.py [hunk 3]
```diff
@@ -414,48 +417,84 @@ def get_filetype(path_name):
...
+                else:
+                    raise Exception
+
                 for intermediate_path in reversed(intermediate_folders):
-                    os.chmod(intermediate_path, mode_intermediate)
+                    if intermediate_mode is not None:
+                        os.chmod(intermediate_path, intermediate_mode)
+                    if intermediate_group is not None:
+                        chgrp_if_not_world_writable(path, intermediate_group)
+                        os.chmod(path, intermediate_mode)  # reset sticky bit
```
Comment (albestro) [inline lib/spack/llnl/util/filesystem.py]: in both lines: you are modifying `path` permissions, but I would say that they should be set on `intermediate_path` instead

File: lib/spack/llnl/util/filesystem.py [hunk 4]
```diff
@@ -414,48 +417,86 @@ def get_filetype(path_name):
...
                 # OS mode will be used
+                if default_perms == 'args':
+                    intermediate_mode = mode
+                    intermediate_group = group
+                elif default_perms == 'parents':
+                    stat_info = os.stat(last_parent)
+                    intermediate_mode = stat_info.st_mode
+                    intermediate_group = stat_info.st_gid
+                else:
+                    raise Exception
```
Comment (tgamblin) [inline lib/spack/llnl/util/filesystem.py]: This should raise a descriptive `ValueError`

**Pull request merged**